pandas/io/formats/csvs.py

```diff
@@ -75,50 +83,92 @@ def __init__(
         self.mode = ioargs.mode
 
+        # GH21227 internal compression is not used for non-binary handles.
+        if (
+            self.compression
+            and hasattr(self.path_or_buf, "write")
+            and "b" not in self.mode
+        ):
+            warnings.warn(
+                "compression has no effect when passing a non-binary object as input.",
+                RuntimeWarning,
+                stacklevel=2,
+            )
+            self.compression = None
+
         self.sep = sep
         self.na_rep = na_rep
         self.float_format = float_format
         self.decimal = decimal
-
         self.header = header
         self.index = index
         self.index_label = index_label
-        if encoding is None:
-            encoding = "utf-8"
-        self.encoding = encoding
         self.errors = errors
-
-        if quoting is None:
-            quoting = csvlib.QUOTE_MINIMAL
-        self.quoting = quoting
-
-        if quoting == csvlib.QUOTE_NONE:
-            # prevents crash in _csv
-            quotechar = None
+        self.quoting = quoting or csvlib.QUOTE_MINIMAL
         self.quotechar = quotechar
-
         self.doublequote = doublequote
         self.escapechar = escapechar
-
         self.line_terminator = line_terminator or os.linesep
-
         self.date_format = date_format
+        self.cols = cols
+        self.chunksize = chunksize
+
+    @property
+    def index_label(self) -> IndexLabel:
+        return self._index_label
+
+    @index_label.setter
+    def index_label(self, index_label: IndexLabel) -> None:
+        if index_label is not False:
+            if index_label is None:
+                index_label = self._get_index_label_from_obj()
+            elif not isinstance(index_label, (list, tuple, np.ndarray, ABCIndexClass)):
+                # given a string for a DF with Index
+                index_label = [index_label]
+        self._index_label = index_label
+
+    def _get_index_label_from_obj(self):
+        if isinstance(self.obj.index, ABCMultiIndex):
+            return self._get_index_label_multiindex()
+        else:
+            return self._get_index_label_flat()
+
+    def _get_index_label_multiindex(self):
+        return [name or "" for name in self.obj.index.names]
 
-        self.has_mi_columns = isinstance(obj.columns, ABCMultiIndex)
+    def _get_index_label_flat(self):
+        index_label = self.obj.index.name
+        return [""] if index_label is None else [index_label]
 
+    @property
+    def quotechar(self):
+        if self.quoting != csvlib.QUOTE_NONE:
+            # prevents crash in _csv
+            return self._quotechar
+
+    @quotechar.setter
+    def quotechar(self, quotechar):
+        self._quotechar = quotechar
+
+    @property
+    def has_mi_columns(self):
+        return isinstance(self.obj.columns, ABCMultiIndex)
+
+    @property
+    def cols(self):
+        return self._cols
+
+    @cols.setter
+    def cols(self, cols):
```

Comment:
and type if you can

Reply:
Added, but with the limitations.
See 1346995